src/main/java/roboy/context/Context.java

```diff
@@ -1,135 +1,197 @@
+
+    /* OBSERVERS DEFINED HERE */
+    private final FaceCoordinatesObserver FACE_COORDINATES_OBSERVER;
+
+
+    // By calling getInstance, Context initializes its updaters and observers.
+    private static Context context = Context.getInstance();
+
+    /**
+     * Builds the class to instance maps.
+     */
     private Context() {
-        FaceCoordinates faceCoordinates = new FaceCoordinates();
-        DialogTopics dialogTopics = new DialogTopics();
-
-        values = new ImmutableClassToInstanceMap.Builder<AbstractValue>()
-                // Collect all Values centrally right here.
-                .put(FaceCoordinates.class, faceCoordinates)
-                .build();
-
-        valueHistories = new ImmutableClassToInstanceMap.Builder<AbstractValueHistory>()
-                // Collect all ValueHistories centrally right here.
-                .put(DialogTopics.class, dialogTopics)
-                .build();
-
-        // Initialize and store the external updaters.
-        externalUpdaters.add(new FaceCoordinatesUpdater(faceCoordinates, 1));
-        // Initialize direct updaters with their targets.
-        internalUpdaters.put(DialogTopicsUpdater.class, new DialogTopicsUpdater(dialogTopics));
+        /* INTERNAL UPDATERS INITIALIZED HERE */
+        DIALOG_TOPICS_UPDATER = new DialogTopicsUpdater(DIALOG_TOPICS.valueHistory);
+        DIALOG_INTENTS_UPDATER = new DialogIntentsUpdater(DIALOG_INTENTS.valueHistory);
+        ACTIVE_INTERLOCUTOR_UPDATER = new ActiveInterlocutorUpdater(ACTIVE_INTERLOCUTOR.value);
+
+        /* OBSERVERS INITIALIZED AND ADDED HERE */
+        FACE_COORDINATES_OBSERVER = new FaceCoordinatesObserver();
+//        FACE_COORDINATES.value.addObserver(FACE_COORDINATES_OBSERVER);
     }
 
+    /**
+     * Starts up the external updaters (which need a ROS main node).
+     * @param ros
+     */
+    public void initializeROS(RosMainNode ros) {
+        Context ctx = Context.getInstance();
+        synchronized (initializationLock) {
+            // Initialize only if not already done.
+            if(!rosInitialized) {
+
+                /* EXTERNAL UPDATERS INITIALIZED HERE */
+//                AUDIO_ANGLES_UPDATER = new AudioDirectionUpdater(AUDIO_ANGLES.valueHistory, ros);
+//                ROS_TEST_UPDATER = new ROSTestUpdater(ROS_TEST.valueHistory, ros);
+                // TODO Add a FACE_COORDINATE_UPDATER.
+                // Edit the data type and integration tests, once the real data type is used.
+
+                rosInitialized = true;
+            }
+        }
+    }
+
+    /**
+     * The access point to Context, including thread-safe Singleton initialization.
+     */
     public static Context getInstance() {
         if (context == null) {
-            context = new Context();
+            // Extra block instead of synchronizing over entire getInstance method.
+            // This way, we do not sync when context was initialized earlier -> better performance.
```

Comment:
Better performance - yes, absolutely.
But what happens if two threads simultaneously enter ``getInstance()`` (this is possible, it's not synchronized), check for ``null`` (assume it is the case) and then create ``new Context()`` twice? Singleton is broken :(

I think there are a few good examples here: https://stackoverflow.com/questions/11165852/java-singleton-and-synchronization

Actually, I wouldn't worry about the performace too much. Here it is not critical since we don't call ``getInstance()`` 10000 times per second.

Reply:
That's why we check for if (context == null) twice, once on line 96 and again on 101, inside the synced block.

So, if two threads call getInstance() before Context was initialized, they compete for the lock object of the synchronized block. Only one will be able to execute the initialization. The other will enter the block after the lock had been released by the other thread, detect that Context is no longer null, and return without re-initialization.

And for most of the runtime, no synchronization in getInstance() is needed. Not much, since it is a short method, but still good practice :)